### PR TITLE
Models SVM/SVR

### PR DESCRIPTION
**Pull request by Klopfe**

### CODE REVIEW

**Comment by agramfort** [comment]
there is quite some work to polish things here. Besides I would use cvxpy and not cvxopt in our tests to avoid adding yet anothet dependency to the package for testing

File: sparse_ho/algo/forward.py [hunk 1]
```diff
@@ -104,14 +104,14 @@ def get_beta_jac_iterdiff(
         alphas = np.ones(n_features) * alpha
     ############################################
     # warm start for beta
-    beta, r = model._init_beta_r(X, y, mask0, dense0)
+    beta, residuals = model._init_beta_residuals(X, y, mask0, dense0)
```
Comment (agramfort) [inline sparse_ho/algo/forward.py:107]: since you call it beta and not betas I would suggest you call it residual (no s)

File: sparse_ho/models/svm.py [hunk 2]
```diff
@@ -23,49 +23,57 @@ class SVM(BaseModel):
...
-            dbeta = np.sum(y * dr * X.T, axis=1)
-        return dbeta, dr
-
-    def _init_beta_r(self, X, y, mask0, dense0):
+            if self.residuals.shape[0] != n_samples:
+                dbeta = np.zeros(n_features)
+            else:
+                dresiduals = self.dresiduals.copy()
+                if issparse(X):
+                    dbeta = (X.T).multiply(y * dresiduals)
```
Comment (agramfort) [inline sparse_ho/models/svm.py:48]:
```suggestion
                    dbeta = X.T.multiply(y * dresiduals)
```

File: sparse_ho/models/svm.py [hunk 3]
```diff
@@ -23,49 +23,57 @@ class SVM(BaseModel):
...
+                    dbeta = np.sum(dbeta, axis=1)
+                    dbeta = np.squeeze(np.array(dbeta))
+                else:
+                    dbeta = np.sum(y * dresiduals * X.T, axis=1)
```
Comment (agramfort) [inline sparse_ho/models/svm.py:52]:
```suggestion
                    dbeta = np.sum((y * dresiduals) * X.T, axis=1)
```

File: sparse_ho/models/svm.py [hunk 4]
```diff
@@ -23,49 +23,57 @@ class SVM(BaseModel):
...
+            if self.residuals.shape[0] != n_samples:
+                beta = np.zeros(n_features)
             else:
-                beta = np.sum(y * r * X.T, axis=1)
-        return beta, r
+                residuals = self.residuals
+                if issparse(X):
+                    beta = (X.T).multiply(y * residuals)
+                    beta = np.sum(beta, axis=1)
+                    beta = np.squeeze(np.array(beta))
```
Comment (agramfort) [inline sparse_ho/models/svm.py:68]:
```suggestion
                    beta = np.ravel(beta)
```
?

File: sparse_ho/models/svm.py [hunk 5]
```diff
@@ -23,49 +23,57 @@ class SVM(BaseModel):
...
-                beta = np.sum(beta, axis=1)
-                beta = np.squeeze(np.array(beta))
+            if self.residuals.shape[0] != n_samples:
+                beta = np.zeros(n_features)
             else:
-                beta = np.sum(y * r * X.T, axis=1)
-        return beta, r
+                residuals = self.residuals
+                if issparse(X):
+                    beta = (X.T).multiply(y * residuals)
```
Comment (agramfort) [inline sparse_ho/models/svm.py:66]:
```suggestion
                    beta = X.T.multiply(y * residuals)
```

File: sparse_ho/models/svm.py [hunk 6]
```diff
@@ -23,49 +23,57 @@ class SVM(BaseModel):
...
+                    beta = np.sum(beta, axis=1)
+                    beta = np.squeeze(np.array(beta))
+                else:
+                    beta = np.sum(y * residuals * X.T, axis=1)
```
Comment (agramfort) [inline sparse_ho/models/svm.py:70]:
```suggestion
                    beta = (y * residuals) * np.sum(X, axis=0)
```
if actually the same this would be much faster and you could precompute np.sum(X, axis=0)

File: sparse_ho/models/svm.py [hunk 7]
```diff
@@ -23,49 +23,57 @@ class SVM(BaseModel):
...
+    estimator: instance of ``sklearn.base.BaseEstimator``
+        An estimator that follows the scikit-learn API.
     """
 
     def __init__(self, estimator=None, max_iter=100):
         self.estimator = estimator
         self.max_iter = max_iter
         self.dual = True
-        self.dr = None
+        self.dresiduals = None
```
Comment (agramfort) [inline sparse_ho/models/svm.py:34]: this attributes do not seem documented in the docstring.

also is residual y - X @ beta even in classification?

is SVM just for regression? could we call it SVC et SVR?

**Review by agramfort — COMMENTED** [review]
thanks a lot @Klopfe !


some more feedback

File: sparse_ho/models/svm.py (hunk 7 above)
Comment (Klopfe) [reply sparse_ho/models/svm.py:34]: it's not the quantity y - X @ beta for SVC / SVR it corresponds to the dual variables. 
Yes I can change the name of the model for SVC for classification
I will add the documentation

Comment (Klopfe) [reply sparse_ho/models/svm.py:34]: is it too misleading to call it residuals for these models ? Up to a scaling factor the dual variables of the Lasso are also the residuals right ?

**Comment by agramfort** [comment]
yes then residual is a bad choice of name

>

File: sparse_ho/models/svm.py (hunk 7 above)
Comment (Klopfe) [reply sparse_ho/models/svm.py:34]: Well we could name it dual for the SVM type models but the problem is that it appears in forward/implicit forward (algorithms) code which has to be generic for all model types. Do you have suggestions?

**Comment by agramfort** [comment]
make me some proposals. How about?

dual_var
aux_var (maybe too generic)

>

File: sparse_ho/models/svm.py (hunk 7 above)
Comment (Klopfe) [reply sparse_ho/models/svm.py:34]: dual_var sounds good to me if it's not too misleading for l1 type models.

**Comment by agramfort** [comment]
let's do this then

>

**Comment by QB3** [comment]
Everything is testes and green, I propose to merge now, and to polish in another PR @Klopfe .
Good for you @agramfort ?

**Comment by agramfort** [comment]
thx @Klopfe !
